### PR TITLE
This proposal describes how we would add a new field to the install-config during installation to set the load balancer type in AWS either to NLB or ELB.

### PR DESCRIPTION
**Pull request by miheer**
This proposal describes how we would add a new field to the install-config during installation to set the load balancer type in AWS either to NLB or ELB.

### CODE REVIEW

File: enhancements/installer/aws-load-balancer-type.md
Comment (sadasu) [inline enhancements/installer/aws-load-balancer-type.md:69]: So, would AWSLoadBalancerType in the Ingress CR  https://github.com/openshift/api/blob/master/operator/v1/types_ingress.go#L462 be updated with the value provided in the new field proposed for the install-config?

File: enhancements/installer/aws-load-balancer-type.md
Comment (barbacbd) [inline enhancements/installer/aws-load-balancer-type.md:220]: This isn't a change correct? The NLB resources should already be destroyed unless this is a difference resource. The Machine API creates an NLB for the cluster and those resources should be destroyed.

File: enhancements/installer/aws-load-balancer-type.md
Comment (barbacbd) [inline enhancements/installer/aws-load-balancer-type.md:104]: This is being added to the AWS specific platform struct. The `aws` at the beginning is not necessary. Same applies for all solutions below.

File: enhancements/installer/aws-load-balancer-type.md
Comment (barbacbd) [inline enhancements/installer/aws-load-balancer-type.md:41]: Reword this, default it used too many times

File: enhancements/installer/aws-load-balancer-type.md
Comment (miheer) [reply enhancements/installer/aws-load-balancer-type.md:104]: @barbacbd ok.  The main question regarding the API is as to where shall we add the LB type. Shall it be the infra cluster object or ingress cluster object ?  https://github.com/openshift/enhancements/pull/1148/commits/661734713e511f79b1c1477a6317b8675d21491e#diff-a222e3e95fd0b523e[…]89abe29e084502fbd3R108

File: enhancements/installer/aws-load-balancer-type.md
Comment (miheer) [inline enhancements/installer/aws-load-balancer-type.md:169]: @barbacbd   The main question regarding the API is as to where shall we add the LB type. Shall it be the infra cluster object or ingress cluster object ?

File: enhancements/installer/aws-load-balancer-type.md
Comment (wking) [inline enhancements/installer/aws-load-balancer-type.md:11]: You probably want [an installer approver][1] here instead of me.

[1]: https://github.com/openshift/installer/blob/master/OWNERS_ALIASES

File: enhancements/installer/aws-load-balancer-type.md
Comment (wking) [inline enhancements/installer/aws-load-balancer-type.md:47]: I'm unclear on the motivation here.  The cluster has a default.  Great.  Users can override the default.  Great.  If users delete the resource that holds their override, the cluster reverts to the default.  Makes sense to me.  Why can't a user that says "hey, I actually wanted that NLB; I regret deleting the IngressController" adjust the replacement IngressController to ask for an NLB?

This enhancement is currently suggesting a harder-to-accidentally-delete location to store an install-time choice, and it seems easier for both the administrator and component maintainers to have fewer moving pieces by having the user say "this IngressController isn't configured how I'd like, so I'll adjust its `spec`".

Comment (miheer) [reply enhancements/installer/aws-load-balancer-type.md:47]: @wking yes but we need the operator to create the ingress controller with the LB type which user/admin desired for default ingress controller. Please check https://issues.redhat.com/browse/NE-799

comment https://issues.redhat.com/browse/NE-799?focusedCommentId=20329711&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-20329711

File: enhancements/installer/aws-load-balancer-type.md
Comment (miheer) [inline enhancements/installer/aws-load-balancer-type.md:191]: @barbacbd @patrickdillon The main question regarding the API is as to where shall we add the LB type. Shall it be the infra cluster object or ingress cluster object ?

File: enhancements/installer/aws-load-balancer-type.md
Comment (miheer) [inline enhancements/installer/aws-load-balancer-type.md:194]: @barbacbd @patrickdillon @jstuever The main question regarding the API is as to where shall we add the LB type. Shall it be the infra cluster object or ingress cluster object ? In this enhancement I have proposed 3 options so which one would the installer prefer ? cc @Miciah

Comment (miheer) [reply enhancements/installer/aws-load-balancer-type.md:194]: I will be using infra object as it seems to have all the cloud related env. So closing this loop as I need to start implementing an urgent epic.

File: enhancements/installer/aws-load-balancer-type.md
Comment (sadasu) [inline enhancements/installer/aws-load-balancer-type.md:140]: As an Installer reviewer, my vote is for Option 1 as it already seems to have Platform specific config in the the PlatformSpec and PlatformStatus types

File: enhancements/installer/aws-load-balancer-type.md
Comment (sadasu) [inline enhancements/installer/aws-load-balancer-type.md:132]: Since this is an optional parameter, will clb/elb continue to be the default values?

File: enhancements/installer/aws-load-balancer-type.md
Comment (sadasu) [reply enhancements/installer/aws-load-balancer-type.md:140]: Based on https://github.com/openshift/installer/pull/6074 and https://github.com/openshift/api/pull/1209, it appears that the implementation is making progress with adding this field to the Infrastructure CR. Could you please update this proposal to reflect that and add Option 2 here to the "Althernatives" section.

File: enhancements/installer/aws-load-balancer-type.md
Comment (miheer) [reply enhancements/installer/aws-load-balancer-type.md:132]: If nothing is set then the ingress operator will create a clb/elb.

File: enhancements/installer/aws-load-balancer-type.md
Comment (miheer) [reply enhancements/installer/aws-load-balancer-type.md]: @sadasu done

**Comment by miheer** [comment]
@sadasu PTAL. I have made the changes you asked for.

File: enhancements/installer/aws-load-balancer-type.md
Comment (rfredette) [inline enhancements/installer/aws-load-balancer-type.md:175]: should be `lbType`

File: enhancements/installer/aws-load-balancer-type.md
Comment (rfredette) [inline enhancements/installer/aws-load-balancer-type.md:116]: change `awsLBType` to `LBType` here

File: enhancements/installer/aws-load-balancer-type.md
Comment (rfredette) [inline enhancements/installer/aws-load-balancer-type.md:201]: should be `lbType`

**Review by rfredette — APPROVED** [review]
There are a few places where the old `awsLBType` name is still in use that I pointed out, but other than those nit picks, this looks good to me

**Comment by miheer** [comment]
@rfredette done

**Comment by miheer** [comment]
@rfredette I have made the changes as requested so can you please lgtm ?

@sadasu  I have made the changes as requested so can you please approve  ?

File: enhancements/installer/aws-load-balancer-type.md
Comment (JoelSpeed) [inline enhancements/installer/aws-load-balancer-type.md:149]: Are you likely to ever want to expand load balancer configuration to handle other options for the NLBs or ELBs? We may want to make this a discriminated union instead

File: enhancements/installer/aws-load-balancer-type.md
Comment (JoelSpeed) [inline enhancements/installer/aws-load-balancer-type.md:147]: Should this be on the infrastructure object? Infrastructure to me implies things around the control plane, so i'd expect the internal and external control plane load balancers to be here, but, if this is a workload load balancer, maybe it's better suited somewhere else?

Comment (miheer) [reply enhancements/installer/aws-load-balancer-type.md:147]: The installer team voted to use infrastructure object.

File: enhancements/installer/aws-load-balancer-type.md
Comment (miheer) [reply enhancements/installer/aws-load-balancer-type.md:149]: no

Comment (JoelSpeed) [reply enhancements/installer/aws-load-balancer-type.md:149]: How can you be certain about that? What are the other relevant Load balancer options that are available from AWS, does the ingress object expose anything?

Comment (miheer) [reply enhancements/installer/aws-load-balancer-type.md:149]: There are timeout options available for classic but we don't set any options for NLB. However we can have something like 

https://github.com/openshift/api/pull/1209#discussion_r936724881

which you mentioned.

ingress:
  loadBalancer:
    defaultType: NLB

File: enhancements/installer/aws-load-balancer-type.md
Comment (miheer) [reply enhancements/installer/aws-load-balancer-type.md:147]: @JoelSpeed based on our discussions it looks like the right place is ingress cluster object and not the infrastructure cluster object. The default ingresscontroller can have either an internal or external LB. However, The API uses a separate load balancer.
As you mentioned ->
"Should this be on the infrastructure object? Infrastructure to me implies things around the control plane, so i'd expect the internal and external control plane load balancers to be here, but, if this is a workload load balancer, maybe it's better suited somewhere else?"
over here https://github.com/openshift/enhancements/pull/1148#discussion_r932289883

I think ingress cluster object seems to be the right place.

I have made the necessary changes. PTAL.

Copying @sadasu from the installler team as she had voted for this field to be added in the infrastructure object.

File: enhancements/installer/aws-load-balancer-type.md
Comment (miheer) [reply enhancements/installer/aws-load-balancer-type.md:149]: BTW there are options like loadBalancerSourceRanges and Scope for AWS LB types.

**Comment by miheer** [comment]
@JoelSpeed i have made the changes as the api changed a lot. PTAL. Thanks in advance!

File: enhancements/installer/aws-load-balancer-type.md
Comment (JoelSpeed) [inline enhancements/installer/aws-load-balancer-type.md:122]: You don't need to include these lines

File: enhancements/installer/aws-load-balancer-type.md
Comment (JoelSpeed) [inline enhancements/installer/aws-load-balancer-type.md:129]: You should explain what happens when this value is omitted

File: enhancements/installer/aws-load-balancer-type.md
Comment (JoelSpeed) [inline enhancements/installer/aws-load-balancer-type.md:134]: > AwsLBType

This is out of date

File: enhancements/installer/aws-load-balancer-type.md
Comment (JoelSpeed) [inline enhancements/installer/aws-load-balancer-type.md:147]: Indentation is off here and on the other structs, please fix for readability

**Review by JoelSpeed — COMMENTED** [review]
Nits but otherwise this is fine from API perspective at least

File: enhancements/installer/aws-load-balancer-type.md
Comment (miheer) [reply enhancements/installer/aws-load-balancer-type.md:134]: Hi @JoelSpeed ! I am using the https://github.com/openshift/api/blob/master/config/v1/types_ingress.go#L145

Comment (JoelSpeed) [reply enhancements/installer/aws-load-balancer-type.md:143]: Right ok

**Comment by JoelSpeed** [comment]
From an API perspective LGTM. At this point I think the remaining approval should be sought from the ingress and installer teams

**Comment by sadasu** [comment]
From the API perspective, I find that there are too many levels of APIs. To be specific, IngressSpec --> LoadBalancer --> IngressPlatformSpec --> AWSIngressSpec --> AWSLBType where "-->" can be replaced with contains. There is merit in keeping these nested APIs simpler. I can see why they are designed this way, but is there a possibility to make this flatter?

**Comment by miheer** [comment]
> From the API perspective, I find that there are too many levels of APIs. To be specific, IngressSpec --> LoadBalancer --> IngressPlatformSpec --> AWSIngressSpec --> AWSLBType where "-->" can be replaced with contains. There is merit in keeping these nested APIs simpler. I can see why they are designed this way, but is there a possibility to make this flatter?

Actually we did have a deep discussion on this in the API PR https://github.com/openshift/api/pull/1209  
We would need the api to be built in such a way that it could handle LB option like Scope which is not platform specific and not LB Type specific and also for platform and LB Type specific settings like connection idle time out and load balancer source range.

To set LBType the api will look like this
Spec.LoadBalancer.Platform.AWS.Type

To set the scope of the Load balancer(in the future). This is not platform or LB Type specific.
Spec.Loadbalancer.Scope

To set the LoadBalancerSourceRange which is LBType specific(in the future)
Spec.LoadBalancer.Platform.AWS.AWSClassicLoadBalancerParamaters.LoadBalancerSourceRange

**Comment by sadasu** [comment]
Can the AWSLBType be changed on day-2? Allowing/disallowing it would happen in the Ingress Controller. Just checking what the expected behavior would be.

**Comment by miheer** [comment]
> Can the AWSLBType be changed on day-2? Allowing/disallowing it would happen in the Ingress Controller. Just checking what the expected behavior would be.

If you change the LBType then default IC will get created with that LBType if someone deletes it.

**Comment by sadasu** [comment]
> > Can the AWSLBType be changed on day-2? Allowing/disallowing it would happen in the Ingress Controller. Just checking what the expected behavior would be.
> 
> If you change the LBType then default IC will get created with that LBType if someone deletes it.

I mean, without the IC getting deleted at all by any external entity. If just the LBType is changed on day-2, just want to understand the expected behavior. Is the LBType fixed per cluster install and cannot be changed on day-2 or if it is allowed what is the use case for that?

**Comment by sadasu** [comment]
> > > > Can the AWSLBType be changed on day-2? Allowing/disallowing it would happen in the Ingress Controller. Just checking what the expected behavior would be.
> > > 
> > > 
> > > If you change the LBType then default IC will get created with that LBType if someone deletes it.
> > 
> > 
> > I mean, without the IC getting deleted at all by any external entity. If just the LBType is changed on day-2, just want to understand the expected behavior. Is the LBType fixed per cluster install and cannot be changed on day-2 or if it is allowed what is the use case for that?
> 
> We will support switch between LBType but that will be via updating the ingress controller CR configuration. The installer is supposed to set the LBtype in the ingress cluster config object. It will remain the same unless someone modifies the ingress cluster config object and deletes the ingress controller.

Thanks! That answers my question.

**Comment by sadasu** [comment]
LGTM from Installer perspective.

**Comment by miheer** [comment]
Can I please get a /approve ?

**Comment by rfredette** [comment]
Looks good from an Ingress perspective as well. I'm not sure if I have permission to approve in this repo, but if I do,

/approve

File: enhancements/installer/aws-load-balancer-type.md
Comment (Miciah) [inline enhancements/installer/aws-load-balancer-type.md:270]:
```suggestion
- The cluster-ingress-operator will check the cluster ingress config object for the lbType and reconcile to create a default ingress controller
```

File: enhancements/installer/aws-load-balancer-type.md
Comment (Miciah) [inline enhancements/installer/aws-load-balancer-type.md:282]:
```suggestion
- Users need to put either "NLB" or "Classic" as input or else the LB creation will fail.

- To mitigate this we need to validate the input.
```
This might be a good place to mention that there are known issues with using NLBs, such as <https://bugzilla.redhat.com/show_bug.cgi?id=2023681>.

File: enhancements/installer/aws-load-balancer-type.md
Comment (Miciah) [inline enhancements/installer/aws-load-balancer-type.md:292]:
```suggestion
- Set the `lbType` field in the `install-config.yaml` either to "NLB" or "Classic" and see if the default ingress controller gets created with the desired LB type.
```

File: enhancements/installer/aws-load-balancer-type.md
Comment (Miciah) [inline enhancements/installer/aws-load-balancer-type.md:293]: Are these going to be E2E tests in openshift/installer or what?

File: enhancements/installer/aws-load-balancer-type.md
Comment (Miciah) [inline enhancements/installer/aws-load-balancer-type.md:314]: Are there really any stages to graduating this feature?  We can probably put "N/A".

File: enhancements/installer/aws-load-balancer-type.md
Comment (Miciah) [inline enhancements/installer/aws-load-balancer-type.md:332]: Maybe this would be a good place to put the possible failure you mentioned in the "Risks and Mitigations" section: 'Users need to put either "NLB" or "Classic" for `lbType` or omit the field or else the install-config will fail validation.'

File: enhancements/installer/aws-load-balancer-type.md
Comment (Miciah) [inline enhancements/installer/aws-load-balancer-type.md:337]: It isn't clear what this procedure is intended to accomplish.  If the goal is to determine how the ingress operator is determining the default LB type, you could also suggest checking the cluster version or ingress operator version to make sure it is new enough to have the enhancement, using `oc version` or `oc get clusteroperators/ingress -o yaml`, and checking the LB type in the cluster ingress config, using  `oc get ingress.config.openshift.io/cluster -o yaml`.  You could also tell the user how to check the LB type that the operator specified on the service, by checking the appropriate annotations in `oc -n openshift-ingress get svc/router-default -o yaml`.

**Review by Miciah — COMMENTED** [review]
I have some suggested clarifications.  In particular, I would like to make sure we are all in agreement regarding (and that no one is surprised by) my suggested clarifications for the goals and non-goals.

File: enhancements/installer/aws-load-balancer-type.md
Comment (miheer) [reply enhancements/installer/aws-load-balancer-type.md:293]: Manual test.

Comment (miheer) [reply enhancements/installer/aws-load-balancer-type.md:293]: BTW I have already written a unit test case in the installer as well.

**Comment by miheer** [comment]
@Miciah as discussed for whether we shall support only default or all ingress controllers i will be sending an email to service delivery.

File: enhancements/installer/aws-load-balancer-type.md
Comment (Miciah) [reply enhancements/installer/aws-load-balancer-type.md:60]: This hasn't been fixed.

File: enhancements/installer/aws-load-balancer-type.md
Comment (Miciah) [reply enhancements/installer/aws-load-balancer-type.md:68]: This hasn't been fixed.

Comment (Miciah) [reply enhancements/installer/aws-load-balancer-type.md:68]: This hasn't been addressed.

File: enhancements/installer/aws-load-balancer-type.md
Comment (Miciah) [reply enhancements/installer/aws-load-balancer-type.md:73]: This hasn't been addressed.

File: enhancements/installer/aws-load-balancer-type.md
Comment (miheer) [reply enhancements/installer/aws-load-balancer-type.md:68]: This enhancement does not change the default load-balancer type, which is Classic ELB.  --Your suggestion looks good.

**Comment by miheer** [comment]
@Miciah PTAL

File: enhancements/installer/aws-load-balancer-type.md
Comment (Miciah) [reply enhancements/installer/aws-load-balancer-type.md]: I suggested documenting disadvantages to using NLBs, such as <https://bugzilla.redhat.com/show_bug.cgi?id=2023681> in order to forestall questions like this: <https://github.com/openshift/installer/pull/6478#discussion_r992741999>.  Any reason not to document this?

File: enhancements/installer/aws-load-balancer-type.md
Comment (Miciah) [inline enhancements/installer/aws-load-balancer-type.md:299]: What does this mean?

File: enhancements/installer/aws-load-balancer-type.md
Comment (Miciah) [inline enhancements/installer/aws-load-balancer-type.md:337]: The wording here is confusing.  It looks like you just copied and pasted my comment: https://github.com/openshift/enhancements/pull/1148#discussion_r979528038

File: enhancements/installer/aws-load-balancer-type.md
Comment (miheer) [reply enhancements/installer/aws-load-balancer-type.md:299]: I will  say N/A

**Comment by miheer** [comment]
@Miciah PTAL

**Comment by miheer** [comment]
@Miciah PTAL

**Comment by Miciah** [comment]
I see no major issues, none of the other reviewers have raised any concerns over my earlier suggestions, and the implementation PRs have already merged.  
/approve
/lgtm

**Pull request merged**
